src/libexpr/eval.cc

```diff
@@ -65,26 +65,18 @@ static char * dupString(const char * s)
 
 // When there's no need to write to the string, we can optimize away empty
 // string allocations.
-// This function handles makeImmutableStringWithLen(null, 0) by returning the
-// empty string.
-static const char * makeImmutableStringWithLen(const char * s, size_t size)
-{
-    char * t;
+// This function handles makeImmutableString(std::string_view()) by returning
+// the empty string.
+static inline const char * makeImmutableString(std::string_view s) {
+    size_t size = s.size();
```

Comment:
why not:

```cpp
const size_t size {s.size()};
auto destination = allocString(size + 1);
memcpy(destination, s.data(), size);
t[size] = '\0';
return t;
```

the size = 0 case is included.

Reply:
@tfc According to the [C++-standard](https://en.cppreference.com/w/cpp/string/byte/memcpy) the behaviour of `memcpy` is undefined, if the source pointer is an invalid or null pointer, even if count is zero. Thus this codes prevents that potential misuse.
Also the empty string does not use a dynamic memory allocation and each invocation of the function returns the same address. I don't know whether other parts of nix rely on this behaviour...

Reply:
you're right, that sounds sensible. thank you for the explanation.